beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -96,27 +96,38 @@ protected SafeFuture<?> doStart() {
               if (checkpoint.isEmpty()) {
                 return;
               }
-              applyBlocks(genesisState, checkpoint.get().getEpochStartSlot(spec));
-            });
-  }
-
-  public void applyBlocks(final BeaconState genesisState, final UInt64 anchorSlot) {
-    Context context = new Context(genesisState, SpecConfig.GENESIS_SLOT.plus(1), anchorSlot);
-
-    final Bytes32 genesisBlockRoot = BeaconBlockHeader.fromState(genesisState).getRoot();
-    storageUpdateChannel
-        .onReconstructedFinalizedState(genesisState, genesisBlockRoot)
-        .thenComposeChecked(__ -> applyNextBlock(context))
-        .finish(
-            error -> {
-              final Throwable rootCause = Throwables.getRootCause(error);
-              if (rootCause instanceof ShuttingDownException
-                  || rootCause instanceof InterruptedException
-                  || rootCause instanceof RejectedExecutionException) {
-                LOG.debug("Shutting down", rootCause);
-              } else {
-                statusLogger.reconstructHistoricalStatesServiceFailedProcess(error);
-              }
+              final UInt64 anchorSlot = checkpoint.get().getEpochStartSlot(spec);
+
+              chainDataClient
+                  .getLatestAvailableFinalizedState(anchorSlot.decrement())
+                  .thenComposeChecked(
+                      latestState -> {
+                        if (latestState.isPresent()) {
+                          final BeaconState state = latestState.get();
+                          return SafeFuture.completedFuture(
+                              new Context(state, state.getSlot().increment(), anchorSlot));
+                        }
+
+                        final Bytes32 genesisBlockRoot =
+                            BeaconBlockHeader.fromState(genesisState).getRoot();
+                        storageUpdateChannel
+                            .onReconstructedFinalizedState(genesisState, genesisBlockRoot)
+                            .ifExceptionGetsHereRaiseABug();
+                        return SafeFuture.completedFuture(
+                            new Context(genesisState, GENESIS_SLOT.increment(), anchorSlot));
```

Comment:
We probably should chain the future from storing the genesis state here rather than just logging failures.  So:
```suggestion
                        return storageUpdateChannel
                            .onReconstructedFinalizedState(genesisState, genesisBlockRoot)
                            .thenApply(__ ->
                            new Context(genesisState, GENESIS_SLOT.increment(), anchorSlot)));
```